### PR TITLE
[WIP] Add request body PipeReader decorator

### PR DESCRIPTION
**Pull request by halter73**
#2340

I'll get some benchmark results as soon as I have a chance.

### CODE REVIEW

File: src/Kestrel.Core/Internal/Http2/Http2MessageBody.cs [hunk 1]
```diff
@@ -16,22 +18,36 @@ protected Http2MessageBody(Http2Stream context)
...
             }
         }
 
-        protected override Task OnConsumeAsync() => Task.CompletedTask;
-
-        public override Task StopAsync()
+        public override void Stop()
         {
-            _context.RequestBodyPipe.Reader.Complete();
             _context.RequestBodyPipe.Writer.Complete();
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http2/Http2MessageBody.cs]: Shouldn't the RequestBodyPipe go away entirely?

Comment (davidfowl) [reply src/Kestrel.Core/Internal/Http2/Http2MessageBody.cs]: Ah http2

File: src/Kestrel.Core/Internal/Http/Http1MessageBody.cs [hunk 2]
```diff
@@ -236,7 +73,9 @@ private void TryStopTimingReads()
                     BadHttpRequestException.Throw(RequestRejectionReason.UpgradeRequestCannotHavePayload);
                 }
 
-                return new ForUpgrade(context);
+                var body = new ForUpgrade(context);
+                context.RequestBodyPipeReader = new Http1MessageBodyPipeReader(context, body);
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http/Http1MessageBody.cs]: This allocates a new PipeReader per request. Ideally we could just reuse this like we reuse the Stream. It should be resettable.

File: src/Kestrel.Core/Internal/Http/Http1MessageBodyPipeReader.cs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+using System.Collections;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
+{
+    public class Http1MessageBodyPipeReader : PipeReader, IAwaiter<ReadResult>
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http/Http1MessageBodyPipeReader.cs]: Where do we throw                      BadHttpRequestException.Throw(RequestRejectionReason.UnexpectedEndOfRequestContent);
 now?

File: src/Kestrel.Core/Internal/Http/Http1MessageBodyPipeReader.cs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+                _lastReadResult = new ReadResult(ReadOnlyBuffer<byte>.Empty, isCancelled: false, isCompleted: false);
+            }
+        }
+
+        private void AdvanceMessageBody(in SequencePosition consumed, in SequencePosition examined)
+        {
+            long consumedBytes;
+
+            // REVIEW: Is it possible for this condition to be false when the entire buffer was consumed?
+            if (consumed == _lastReadResult.Buffer.End)
```
Comment (davidfowl) [inline src/Kestrel.Core/Internal/Http/Http1MessageBodyPipeReader.cs]: @pakrym is this comparison still safe?

**Comment by pakrym** [comment]
When are we planning to expose body pipe?

**Comment by muratg** [comment]
We decided we're not doing this in PR.

**Pull request closed without merging**